oauth2client/contrib/gce.py

```diff
@@ -86,31 +132,69 @@ class AppAssertionCredentials(AssertionCredentials):
     """
 
     @util.positional(2)
-    def __init__(self, scope='', **kwargs):
+    def __init__(self,
+                 scope='',
+                 service_account_email='default',
+                 service_account_info=None,
+                 **kwargs):
         """Constructor for AppAssertionCredentials
 
         Args:
-            scope: string or iterable of strings, scope(s) of the credentials
-                   being requested. Using this argument will have no effect on
-                   the actual scopes for tokens requested. These scopes are
-                   set at VM instance creation time and won't change.
+            scope:
+                string or iterable of strings, scope(s) of the credentials
+                being requested. Using this argument will have no effect on
+                the actual scopes for tokens requested. These scopes are
+                set at VM instance creation time and won't change.
+            service_account_email:
+                the email for these credentials. This can be used with custom
+                service accounts, or left blank to use the default service
+                account for the instance. Usually the compute engine service
+                account.
+            service_account_info:
+                Deserialized JSON object, returned by self.service_account_info
         """
         if scope:
             warnings.warn(_SCOPES_WARNING)
-        # This is just provided for backwards compatibility, but is not
-        # used by this class.
-        self.scope = util.scopes_to_string(scope)
+
+        self._service_account_info = service_account_info or {
+            'email': service_account_email
+        }
+        self._project_id = None
+
         self.kwargs = kwargs
 
         # Assertion type is no longer used, but still in the
         # parent class signature.
         super(AppAssertionCredentials, self).__init__(None)
-        self._service_account_email = None
 
-    @classmethod
-    def from_json(cls, json_data):
-        data = json.loads(_from_bytes(json_data))
-        return AppAssertionCredentials(data['scope'])
+    @property
+    def service_account_info(self):
+        if self._service_account_info.get('email', 'default') == 'default':
+            self._service_account_info = _get_service_account_info(
+                self._service_account_info.get('email', 'default')
+            )
+        return self._service_account_info
+
+    @property
+    def scopes(self):
+        return self._retrieve_scopes()
+
+    @property
+    def service_account_email(self):
+        return self.service_account_info['email']
+
+    @property
+    def project_id(self):
+        if not self._project_id:
+            self._project_id = _get_metadata('project', 'project-id')
+        return self._project_id
+
+    @property
+    def serialization_data(self):
+        return self.service_account_info
```

Comment:
@dhermes @jonparrott 
Should this:
(a) serialize using the external property, meaning if a metadata server request for info hasn't been made, it will be made at serialization time, or 
(b) serialize the internal property, potentially meaning that it could be serialized with the email as 'default' which ambiguously maps to actual service accounts. 

I went with (a), but am open to other opinions